src/Plugin/RulesAction/SystemMailToUsersOfRole.php

```diff
@@ -0,0 +1,179 @@
+ *       description = @Translation("The from e-mail address."),
+ *       required = FALSE
+ *     )
+ *   }
+ * )
+ *
+ * @todo: Add access callback information from Drupal 7.
+ */
+class SystemMailToUsersOfRole extends RulesActionBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The logger channel the action will write log messages to.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * Mail manager.
+   *
+   * @var \Drupal\Core\Mail\MailManagerInterface
+   */
+  protected $mailManager;
+
+  /**
+   * The site configuration.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  protected $config;
+
+  /**
+   * The user storage service.
+   *
+   * @var \Drupal\user\UserStorage
+   */
+  protected $userStorage;
+
+  /**
+   * Constructs a SendMailToUsersOfRole object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The storage service.
+   * @param \Drupal\Core\Mail\MailManagerInterface $mail_manager
+   *   The mail manager service.
+   * @param \Drupal\Core\Config\ImmutableConfig $site_config
+   *   The site configuration.
+   * @param \Drupal\user\UserStorageInterface $user_storage
+   *   The user storage service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, LoggerInterface $logger, MailManagerInterface $mail_manager, ImmutableConfig $site_config, UserStorageInterface $user_storage) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->logger = $logger;
+    $this->mailManager = $mail_manager;
+    $this->config = $site_config;
+    $this->userStorage = $user_storage;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('logger.factory')->get('rules'),
+      $container->get('plugin.manager.mail'),
+      $container->get('config.factory')->get('system.site'),
+      $container->get('entity_type.manager')->getStorage('user')
+    );
+  }
+
+  /**
+   * Send a mail to users of a role.
+   *
+   * @param \Drupal\user\RoleInterface[] $roles
+   *   Array of UserRoles to assign.
+   * @param string $subject
+   *   Subject of the email.
+   * @param string $body
+   *   Email message text.
+   * @param string $from
+   *   Email from address.
+   */
+  protected function doExecute(array $roles, $subject, $body, $from) {
+    // Get the role ids and remove the empty ones (in case for example the role
+    // has been removed in the meantime).
+    $rids = array_filter(array_map(function ($role) {
```

Comment:
The roles property can be either an array of entities or an array of entity ids.
Currently we are addressing it as both.

I prefer to use it as an array of ids.